### PR TITLE
Updated _eval_is_odd to handle more complex inputs

### PR DESCRIPTION
**Pull request by mikaylazgrace**
Changed the function _eval_is_odd to handle integer inputs that have a denominator, such as 
```
n = Symbol('n',integer=True,even=True)
m = Symbol('m',integer=true,even=True)
x = Mul(n,m,S.Half)
```
The example expression x is recognized by SymPy as an integer, but can be decomposed into n,m,and 1/2.  My new function evaluates the oddness of each part and uses this to calculate the oddness of the entire integer.

Addresses issue #8648

### CODE REVIEW

**Comment by asmeurer** [comment]
Can you add some tests (based on the other tests, it looks like `sympy/core/tests/test_arit.py` is the proper file).

File: sympy/core/mul.py [hunk 1]
```diff
@@ -1264,26 +1266,38 @@ def _eval_is_odd(self):
...
+                if t == S.Half:
+                    count_denom += 1
+                elif isinstance(t,Symbol):
+                    if t.is_integer:
+                        if t.is_even:
+                            count_num += 1
+                        elif t.is_odd is None or t.is_even is None:
+                            return None
+                                               
+                elif isinstance(t,int):
```
Comment (asmeurer) [inline sympy/core/mul.py]: This shouldn't be necessary. All elements of args should be SymPy types.

File: sympy/core/mul.py [hunk 2]
```diff
@@ -1264,26 +1266,38 @@ def _eval_is_odd(self):
...
-                        pass
-                    elif acc != 1 and (acc + t).is_odd:
-                        r = False
-                    elif t.is_odd is None:
-                        r = None
-                acc = t
-            return r
+                if t == S.Half:
+                    count_denom += 1
+                elif isinstance(t,Symbol):
```
Comment (asmeurer) [inline sympy/core/mul.py]: Why this restriction? Wouldn't the logic work for any expression?

File: sympy/core/mul.py [hunk 3]
```diff
@@ -1264,26 +1266,38 @@ def _eval_is_odd(self):
...
         if is_integer:
-            r, acc = True, 1
+            #The entire expression is an integer, but it
+            #can be decomposed into fractions (i.e.
+            #numerator and denominator.
+            #The counters keep track of multiples of 2 in
+            #numerator and denominator to determine oddness.
+            count_denom = 0
+            count_num = 0
             for t in self.args:
```
Comment (asmeurer) [inline sympy/core/mul.py]: You can use the `fraction` function to get the numerator and denominator of an expression.

File: sympy/core/mul.py (hunk 2 above)
Comment (mikaylazgrace) [inline sympy/core/mul.py]: Are you referring to the restriction regarding the S.Half?  I'm only keeping track of factors of 2 in the numerator and denominator, since we know the whole expression is an integer.  But I see what you're saying, like if someone input an expression with a 1/6, 1/6 has a factor of 2 in the denominator but with how it handles right now, this factor of 2 would not be recognized by the if statement.  I could change it to also include any denominator that is even, so 0 mod 2.  Would this fix what you were referring to?

File: sympy/core/mul.py (hunk 3 above)
Comment (mikaylazgrace) [inline sympy/core/mul.py]: Okay, good to know!  I wasn't aware of that function.  I'll review and see how that could be implemented to make the function clearer.

File: sympy/core/mul.py (hunk 2 above)
Comment (asmeurer) [inline sympy/core/mul.py]: I guess I was thinking of symbols in the denominator, but in that case you don't even know if the expression is an integer.

Comment (mikaylazgrace) [inline sympy/core/mul.py]: I thought the first if statement should handle that for the entire expression.  It will only enter into the other loop if is_integer  is True for the entire expression.  Then it moves on to evaluate the decomposition of the expression.  Does this logic work?

File: sympy/core/mul.py (hunk 1 above)
Comment (mikaylazgrace) [inline sympy/core/mul.py]: Can you clarify what you mean by this?  My reasoning behind this piece is that if t isn't a fraction (i.e. is_integer) then that means it's part of the numerator.  If it's even (has, at the minimum, a factor of 2 in it), then we know there is a 2 in the numerator, so the count_num +=1.  For the next elif, if any part of the expression is unknown to be odd or even, then we can't evaluate it, so None.

For the bottom elif isinstance, is int not a SymPy type?  Should it be Integer instead?
Hope I explained my reasoning a bit!  Let me know if I should clarify more or any other suggestions.

File: sympy/core/mul.py (hunk 2 above)
Comment (asmeurer) [inline sympy/core/mul.py]: I mean like `x/y`, which is a `Mul` in SymPy (it's represented as `Mul(x, Pow(y, -1))`.  If you knew it was an integer you could say something, but as I said you don't in general.

For the numeric part, you should check if the denominator is even.

File: sympy/core/mul.py (hunk 1 above)
Comment (asmeurer) [inline sympy/core/mul.py]: Yes, `int` is the Python type, but the SymPy type is `Integer` (and similarly `Rational` for rationals).

```
>>> (2*x).args
(2, x)
>>> type((2*x).args[0])
<class 'sympy.core.numbers.Integer'>
```

Comment (asmeurer) [inline sympy/core/mul.py]: And `Integer` is not a subtype of `int`, so this code will never be called.

File: sympy/core/mul.py (hunk 2 above)
Comment (asmeurer) [inline sympy/core/mul.py]: Actually, I suppose more general logic could try to find the largest power of 2 that divides the denominator, and use that (I suppose you already imagined this logic with having "count_denom"). There's probably a SymPy function that returns this but I'm not sure what it is (maybe `factorint(n, limit=3).get(2, 0)`).

Also, a fact worth knowing for this logic is that a rational number can only appear once in a Mul, and it will always be the first arg. A helpful method here is `self.as_coeff_mul()`, which returns `(coeff, args)` where coeff is the numeric coefficient and `args` is the rest:

```py
>>> (2*x*y).as_coeff_mul()
(2, (x, y))
```

Also, rational numbers are always reduced to lowest terms (as you probably guessed).

File: sympy/core/mul.py (hunk 1 above)
Comment (mikaylazgrace) [inline sympy/core/mul.py]: Of course, I should have figured this out anyways since everything else is called .is_integer and so forth.  That one's an easy fix at least!

File: sympy/core/mul.py (hunk 2 above)
Comment (mikaylazgrace) [inline sympy/core/mul.py]: I didn't know about '''self.as_coeff_mul()''', so I'll use this method as it would make it easier to evaluate.    I think the counters make sense as a way to keep track of the powers of 2 in the numerator, so I think I will keep that logic.  I'll try to find the function you were talking about regarding factoring the denominator.  Thanks again, you've been amazingly helpful!

**Comment by mikaylazgrace** [comment]
@asmeurer , [here](https://github.com/mikaylazgrace/sympy/blob/Working-on-issue-8648/sympy/core/mul.py) is the updated work (line 1266) incorporating some of the ideas we discussed above.  I'm unsure if I am calling the functions .fraction() and .as_coeff_mul() correctly.

An error raised by Travis in the original PR said that when I called the functions (around line 90 in mul.py), "self is not defined".  Any ideas on how to call the functions?  Below is how I did it (which didn't work):

```
class Mul(Expr, AssocOp):

    __slots__ = []

    is_Mul = True
    is_odd = self._eval_is_odd()
    is_even = self._eval_is_even()
```

**Comment by asmeurer** [comment]
self is only defined inside of methods (where self is the first argument). There's no need to define is_odd or is_even. SymPy defines those automatically from _eval_is_odd and _eval_is_even.

File: sympy/core/mul.py [hunk 4]
```diff
@@ -1258,32 +1261,72 @@ def _eval_pos_neg(self, sign):
     def _eval_is_negative(self):
         if self.args[0] == -1:
             return (-self).is_positive  # remove -1
-        return self._eval_pos_neg(-1)
-
+        return self._eval_pos_neg(-1)     
+    
     def _eval_is_odd(self):
+ 
+  """ First we check to see if the expression as a whole is an integer because
```
Comment (asmeurer) [inline sympy/core/mul.py]: Indent this so that it matches the `def` line.

File: sympy/core/mul.py [hunk 5]
```diff
@@ -1258,32 +1261,72 @@ def _eval_pos_neg(self, sign):
...
-                    elif t.is_odd is None:
-                        r = None
-                acc = t
-            return r
-
-        # !integer -> !odd
-        elif is_integer is False:
-            return False
+            if is_integer:
+                self.as_coeff_mul()
```
Comment (asmeurer) [inline sympy/core/mul.py]: You aren't using the result of this function.

Comment (mikaylazgrace) [inline sympy/core/mul.py]: I think I fixed that!  When I was trying to test it, the error below happened which doesn't look related to my function, but prevented me from testing:
   ```
 from .simplify import (simplify, hypersimp, hypersimilar,
  File "/home/princess_m/Documents/sympy_patch/sympy/sympy/simplify/simplify.py", line 5, in <module>
    from sympy.core import (Basic, S, Add, Mul, Pow,
ImportError: cannot import name Add

```

Comment (asmeurer) [inline sympy/core/mul.py]: You need to move the imports that you added at the top of the file to inside this function. Otherwise, you get circular imports (which is the error you're seeing).

Comment (mikaylazgrace) [inline sympy/core/mul.py]: Okay, got it!  That was definitely confusing me, as I didn't know if it was related to my function or not.

File: sympy/core/mul.py (hunk 4 above)
Comment (mikaylazgrace) [inline sympy/core/mul.py]: Thanks, I fixed that.  This is relatively minor, but is it okay that it's a docstring with """ and not commented with #?  I know a lot of programs can be strict about what should be a docstring and what is just regular commenting.

Comment (asmeurer) [inline sympy/core/mul.py]: A comment would be better. It doesn't matter too much since this is a private method, but in general this sort of thing would go in a comment.

Comment (mikaylazgrace) [inline sympy/core/mul.py]: Okay, I'll change that to comments.  Another thing I changed was in response to an error Travis was giving me that said: `AttributeError: 'NegativeOne' object has no attribute 'fraction'`.  I believe I fixed that with:
```
if coeff == 1 or coeff == -1:  
    r = True  
    for t in symbols.arg[1]:  
        if t.is_even:  

```

Comment (mikaylazgrace) [inline sympy/core/mul.py]: Okay, I'll change that to comments.  Another thing I changed was in response to an error Travis was giving me that said: `AttributeError: 'NegativeOne' object has no attribute 'fraction'`.  I believe I fixed that with:
```
if coeff == 1 or coeff == -1:  
    r = True  
    for t in symbols.arg[1]:  
        if t.is_even:  

```

File: sympy/core/mul.py [hunk 6]
```diff
@@ -1258,32 +1258,74 @@ def _eval_pos_neg(self, sign):
...
+                    if t.is_even:
+                        return False
+                    elif t.is_even == None:
                         r = None
-                acc = t
-            return r
+                return r
+                
+            else:
+                (numer,denom) = coeff.fraction()
```
Comment (asmeurer) [inline sympy/core/mul.py]: I don't think this works for any object, because there is no `fraction` method. It should be `fraction(coeff)`, or, better, `coeff.as_numer_denom()`.

Comment (mikaylazgrace) [inline sympy/core/mul.py]: Sure, why do you think coeff.as_numer_denom() is better than fraction(coeff) in this case?

Comment (asmeurer) [inline sympy/core/mul.py]: I think it should be faster, since coeff is known to be a rational number (though I haven't tested it).

Comment (mikaylazgrace) [inline sympy/core/mul.py]: Okay, thank you, good to know!

**Comment by mikaylazgrace** [comment]
@asmeurer , Travis is giving me this error:       
```
File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/core/mul.py", line 1291, in _eval_is_odd
        symbols = list(chain.from_iterable(symbols)) #.as_coeff_mul() returns a tuple for arg[1] so we change it to a list
    TypeError: 'exp_polar' object is not iterable
```

I'm a bit unsure how to approach/fix this.  The added symbols = list... was in response to the fact that .as_coeff_mul() returns a tuple in arg[1] which isn't iterable.

**Comment by asmeurer** [comment]
I'm not clear what you are trying to do on that line. You should be able to just have 

```
coeff, args = self.as_coeff_mul()
for arg in args:
    ...
```

**Comment by mikaylazgrace** [comment]
I had some equivalent of that, but Travis said that I was trying to iterate over a tuple. I thought .as_coeff_mul() returned a tuple for the second part? Maybe I'm incorrect there.

**Comment by mikaylazgrace** [comment]
I just made a small change that might solve it:  I'm looping over symbols instead of symbols.arg[1].  

I also changed .as_coeff_mul() per your suggestion (around line 1288):
```
        if is_integer:
            coeff,symbols = self.as_coeff_mul()
            if coeff == 1 or coeff == -1:
                r = True
                for t in symbols:
```

File: sympy/core/mul.py [hunk 7]
```diff
@@ -1258,32 +1258,73 @@ def _eval_pos_neg(self, sign):
...
-                    return None
-                elif t.is_even:
-                    r = False
-                elif t.is_integer:
-                    if r is False:
-                        pass
-                    elif acc != 1 and (acc + t).is_odd:
-                        r = False
-                    elif t.is_odd is None:
+            coeff,symbols = self.as_coeff_mul()
```
Comment (asmeurer) [inline sympy/core/mul.py]: `args` would be a better name here. The arguments are not necessarily Symbol objects.

File: sympy/core/mul.py [hunk 8]
```diff
@@ -1258,32 +1258,73 @@ def _eval_pos_neg(self, sign):
...
+                
+            else:
+                (numer,denom) = fraction(coeff)
+                if numer % 2 == 0:
+                    return False
 
-        # !integer -> !odd
-        elif is_integer is False:
+                elif denom % 2 == 0:
+                    factordict = denom.factorint()
```
Comment (asmeurer) [inline sympy/core/mul.py]: You need to set the limit argument to factorint, so that it doesn't try to find very large prime factors.

Comment (mikaylazgrace) [inline sympy/core/mul.py]: What do you mean by the limit argument?

Comment (asmeurer) [inline sympy/core/mul.py]: factorint has a limit argument. See http://docs.sympy.org/latest/modules/ntheory.html#sympy.ntheory.factor_.factorint

Comment (asmeurer) [inline sympy/core/mul.py]: Actually, it looks like it's better to use `denom.factors` with the limit argument, which disables expensive factoring algorithms automatically.

Comment (mikaylazgrace) [inline sympy/core/mul.py]: I changed it to `factordict = Factors(denom)` based on the example in sympy/core/exprtools.py:
```
        >>> from sympy.core.exprtools import Factors
        >>> from sympy.abc import x
        >>> from sympy import I
        >>> e = 2*x**3
        >>> Factors(e)
```

Is this correct or should it be` denom.factors`?

**Comment by czgdp1807** [comment]
@mikaylazgrace Are you still interested to work on it?

**Comment by sachin-4099** [comment]
I will takeover and come up with a PR soon.
